### ISSUE TITLE
Allow custom decoder function

### ISSUE THREAD

**Issue by laczoka** [dakrone/cheshire#49]
Hi,

I find myself often parsing a json string into a clj datastructure, then do a deep walk to coerce values into a custom type based on their key.

I would love to be able to do pass an (fn [k v] ...) function which does the coercion of values during the decoding.

Wdyt?

**Comment by jzwolak** [comment]
I might agree with @dakrone , except that Chesire converts Clojure keywords to strings when creating JSON and then leaves them as strings when converting back to Clojure data structures. This is problematic as the encode and decode are not reverse operations of each other. Most users will assume they are. The problem, of course, is that JSON doesn't have an equivalent of Clojure's keyword type, so data is lost when converting to JSON. Providing the user with a way to handle this as part of the encode/decode fns would allow the user to make them the reverse operations for the user's data. Write the data to a web app backend... read it later... and something else comes out because of this.

So basically, Chesire makes a decision about type conversion that limits what data can be conveniently used and provides no mechanism for choosing otherwise. Yet, the key-fn is provided for the keys. I don't know... I think it's worth doing this.
